### ISSUE TITLE
kvstreamer: improve the stability

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#82160]
Currently, the usage of the streamer library seems to make things less stable memory-wise than before - which goes against the goal of the library. We should achieve higher number in `tpch_concurrency` roachtest with the streamer enabled. I think the root cause is that the streamer creates many more objects in general (increasing pressure on the GC), and large slices are not accounted for yet.

We should:
- remove allocations as much as possible
- make sure all things are accounted for
- ~consider making `Result` an interface so that depending on `Hints` and `Mode` we could have smaller-size implementations.~

Jira issue: CRDB-16320

**Comment by yuzefovich** [comment]
With 10 runs of `tpch_concurrency` test (on master with #82384, #82387, #82865, #83010 cherry-picked) for the streamer OFF I got `78.5` and for the streamer ON I got `84.9`, on average.

**Comment by yuzefovich** [comment]
I did 10 runs of different `tpch_concurrency` configs on a binary that included #84230 and #84319 (as well as already merged #84229), and here is what I got (the higher, the better):
- `high_refresh_spans_bytes`: 78.5
- `no streamer`: 107.6
- `with streamer`: 104.

At this point, it seems like the stability of the streamer is roughly on par with the non-streamer config, so I'm marking this issue as "done", and will make sure to push #84230 and #84319 over the finish line. I also might look into the stability aspect during the stability period.
